src/svc_auth_gss.c

```diff
@@ -493,6 +493,7 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 			offset = 0;
 		} else if (offset >= gd->win || (gd->seqmask & (1 << offset))) {
 			*no_dispatch = true;
+			rc = RPCSEC_GSS_CREDPROBLEM;
```

Comment:
If I go back to 1.5, I see that this case was explicitly returning AUTH_OK (which is what it's returning now).  I don't understand this code that well, could you explain why this is an error rather than OK?

Reply:
Two of our customers have reported below crash:

(gdb) bt
#0  0x00003fff7ad85328 in raise () from /lib64/libpthread.so.0
#1  0x0000000010071824 in crash_handler (signo=11, info=0x3ffad870d7c8, ctx=0x3ffad870ca50)
    at /usr/src/debug/nfs-ganesha-2.5.3-ibm036.04-0.1.1-Source/MainNFSD/nfs_init.c:225
#2  <signal handler called>
#3  0x0000000000000024 in ?? ()
#4  0x00003fff7acfad24 in svc_vc_getargs (req=0x3ffb50124ba8, xdr_args=0x1006203c <xdr_COMPOUND4args>, args_ptr=0x3ffb501253b0,
    u_data=0x3ffb501253a8) at /usr/src/debug/nfs-ganesha-2.5.3-ibm036.04-0.1.1-Source/libntirpc/src/svc_vc.c:837
#5  0x000000001006f7b0 in thr_decode_rpc_request (context=0x0, xprt=0x3ffcac1bbca0)
    at /usr/src/debug/nfs-ganesha-2.5.3-ibm036.04-0.1.1-Source/MainNFSD/nfs_rpc_dispatcher_thread.c:2015
#6  0x000000001006fcd0 in thr_decode_rpc_requests (thr_ctx=0x3ffed40048f0)
    at /usr/src/debug/nfs-ganesha-2.5.3-ibm036.04-0.1.1-Source/MainNFSD/nfs_rpc_dispatcher_thread.c:2104
#7  0x000000001016958c in fridgethr_start_routine (arg=0x3ffed40048f0)
    at /usr/src/debug/nfs-ganesha-2.5.3-ibm036.04-0.1.1-Source/support/fridgethr.c:550
#8  0x00003fff7ad78b94 in start_thread () from /lib64/libpthread.so.0
#9  0x00003fff7abb85f4 in clone () from /lib64/libc.so.6

The svc_ah_unwrap is not set which is causing the crash.

After svc_auth_authenticate() is called "no_dispatch" is true and AUTH_OK is returned.
But reqdata->r_u.req.svc.rq_msg.RPCM_ack.ar_verf.oa_flavor is not set.

So this message should not be processed....but it is and causing the crash.

svc_auth_authenticate() header indicates 
"If AUTH_OK then the routine also does the following things:
 * set req->rq_msg.RPCM_ack.ar_verf to the appropriate response verifier;
 * set req->rq_msg.rq_cred_body to the "cooked" form of the credentials"

which is not happening for this scenario...so either we return RPCSEC_GSS_CREDPROBLEM
OR do something like below changes in Ganesha code (as no_dispatch is true):

https://github.com/ntrishal/nfs-ganesha/commit/052fdbfb11e645e34ac2f717e9e7b1f1a138fdea

Reply:
But with this fix, krb5, krb5s, and krb5p all still work?  If so, I can agree.

Reply:
Looking at the KRB5 RFC rfc2203.txt, we do need to drop requests when the sequence number is outside of the window to avoid DOS attacks. Pynfs specifically tests for this. Setting AUTH_OK and no_dispatch is the correct thing to do. The caller, in Ganesha code, should handle it correctly. Let us fix the ganesha project code than libntirpc code here.

Reply:
Okay, so should we close this?

Reply:
Yes we can close this...as fixing in Ganesha code.